### DISCUSSION TITLE
...must be of the type array, bool returned

### DISCUSSION THREAD

**Discussion by franrata** [symfony/symfony#50543] [Q&A]
Hello. First time asking for help, usually i find a solution online.

I get this error:
**Return value of App\Repository\DeviceStatsRepository::getLastStats() must be of the type array, bool returned**

Symfony version 6.

**But If i do dd($resultSet->fetchAssociative());**
i got an array: 
^ array:7 [▼
  "id" => "41939"
  "device_id" => "1"
  "name" => "temp"
  "value" => "20.60"
  "epoch" => "1685108001"
  "timestamp" => "2023-05-26 16:33:35"
  "ip" => "xxx"
]
**and if i dont interrupt the script it shows the bool error.**
im confused, maybe something i just dont see.
Thanks.

Repository function:
```php
    public function getLastStats(int $device, string $name): array
    {
        $conn = $this->getEntityManager()->getConnection();

        $sql = '
            SELECT * FROM device_stats ds
            WHERE ds.name = :name AND ds.device_id = :device
            ORDER BY ds.id DESC
            LIMIT 1
            ';
        $stmt = $conn->prepare($sql);
        $resultSet = $stmt->executeQuery(['device' => $device, 'name' => $name]);

        // returns an array of arrays (i.e. a raw data set)
        //dd($resultSet->fetchAssociative());
        return $resultSet->fetchAssociative();

    }
```

**Comment by ghost** [comment]
Please pay attention how you format the code blocks. The quality of the report reflects respect for the people you are asking for help.

**Reply by ghost** [reply]
Probably it's not allowed to fetch from the ResultSet twice. I've never done that but that's what I suspect happens..

**Reply by franrata** [reply]
thanks for editing, looks way better now.
Where do you say is calling it twice?

**Reply by ghost** [reply]
To be honest it's hard to understand what exactly is the code that causes your issue. The code block shows something else than what's described. The code shows that dd is commented out but above you say `But If i do dd($resultSet->fetchAssociative()); i got an array ... and if i dont interrupt the script it shows the bool error.` - what I understand is that dd is not commented out.
Please be more meaningful.



> Probably it's not allowed to fetch from the ResultSet twice

Looked at the dbal source code and fetchAssociative returns false when there are no more rows to fetch from the database.
Please investigate your problem once again and make sure you don't call `$result->fetchAssociative()` more times that the number of rows in the database

**Reply by franrata** [reply]
When is NOT commented, i got the array.
When IT IS commented i got the error: Return value of App\Repository\DeviceStatsRepository::getLastStats() must be of the type array, bool returned

it is never called twice.

**Comment by MatTheCat** [comment]
You may be missing another invocation of `getLastStats` because `dd` stops the script execution. What if you replace it with `dump`?

**Reply by franrata** [reply]
if i put dump i get both, the array and the same error.

![image](https://github.com/symfony/symfony/assets/4423277/cf406ffa-d7f7-4735-9fae-13e864538035)

I dont understand why responses here are about the DD... thats just to show that the response is an array,... not part of the code.



Why do i get "must be of the type array, bool returned"? thats the question.

**Reply by MatTheCat** [reply]
Because the second call to `getLastStats` yields no result and thus returns `false`, which conflicts with your `array` return type declaration.

`dd` **stops** the script execution, so when you use it `getLastStats` is only called **once**, whereas the problem lies within its **second** invocation.

**Reply by franrata** [reply]
but if i remove the dd i get the error anyway.

**Reply by MatTheCat** [reply]
Please carefully read my previous message.

**Reply by franrata** [reply]
sorry, i dont get it, where is the "second invocation"?

**Reply by MatTheCat** [reply]
Sorry I misunderstood :sweat_smile: 

Could you paste the stack trace you get when the code crashes?

**Comment by franrata** [comment]
no man, this makes no sense, so i change fetchAssociative for fetchAllAssociative
and add the zero index to the result. 

no answer for the question asked.
thanks for trying to help... i dont understand why both replies where fixated on the dd... maybe my bad explaining.
gl.
